### ISSUE TITLE
ReLU uses more VRAM for backward-pass than needed

### ISSUE THREAD

**Issue by MengeTM** [pytorch/pytorch#63027]
## 🚀 Feature

Storing the ReLU output Tensor for backward-pass instead of the input Tensor.

Both the input- and output-Tensor should be equivalent regarding the gradient calculation for ReLUs. However, the output-Tensor is usually already stored for calculation of the backward-pass for the following module, giving an opportunity to reduce VRAM allocated by the GPU.

## Motivation

Large models use a lot of VRAM, often reach the limit of a GPU, and ReLU activation functions may be used several times within a network, resulting in a large memory saving potential.

## Pitch

1. Instead of saving the input Tensor for the torch.nn.ReLU backward-pass, the `output = th.relu(input)` of the module may be saved for the backward-pass.
2. During the backward-pass, the input Tensor is replaced by the output Tensor, e.g. `grad *= output>0`, or however this is realized in the PyTorch code.

This should be applicable due to output-values being <=0 for both the input and the output Tensors.

This saves VRAM, due to ReLU outputs being often an input to following modules, where the input is usually saved for backward-pass gradient calculations. Thus, the output Tensor (ReLU) has only to be saved once, whereas the input Tensor (ReLU) needs to be saved twice.

## Alternatives

Leaving torch.nn.ReLU as is.

## Additional context

The following script repeatedly applies the same Conv2d module and either the new ReLU or torch.nn.ReLU in a closed-loop set-up, and calculates the gradients during a backward-pass.

The script shows an improvement for ReLU regarding VRAM usage (without model parameter, input, label), and an exemplary equivalence of gradients for both versions.

However, the forward+backward pass duration is dependent on which Module is evaluated first, resulting in the same runtime when running the new ReLU first, and in a faster runtime for the new ReLU when running torch.nn.ReLU first.

```
import torch as th
from torch import nn
from torch.autograd import Function

import time

ITERATION = 10
DEVICE = 0
SHAPE = [1, 256, 512, 512]


class ReluFunction(Function):
    @staticmethod
    def forward(ctx, input: th.Tensor) -> th.Tensor:
        relu = th.relu(input)
        ctx.save_for_backward(relu)
        return relu

    @staticmethod
    def backward(ctx, grad: th.Tensor):
        relu, = ctx.saved_tensors
        grad *= (relu > 0)
        return grad


class ReLU(nn.Module):
    def __init__(self):
        super(ReLU, self).__init__()

    def forward(self, input: th.Tensor) -> th.Tensor:
        return ReluFunction.apply(input)


def speed_test(model: nn.Module, module: nn.Module, input: th.Tensor, output: th.Tensor, criterion, iteration=10):
    device = input.device
    module = module.to(device=device)

    input.grad = None
    input.requires_grad_()

    model = model.to(device=device)

    th.cuda.synchronize(device)
    th.cuda.reset_peak_memory_stats(device)
    vram_0 = th.cuda.max_memory_allocated(device)
    time_0 = time.time()

    prediction = input
    for _ in range(iteration):
        prediction = model(prediction)
        prediction = module(prediction)

    th.cuda.synchronize(device)
    vram_forward = th.cuda.max_memory_allocated(device)
    loss = criterion(prediction, output)
    loss.backward()

    th.cuda.synchronize(device)
    print(f'Forward + Backward: {time.time() - time_0:1.2f}s, \n'
          f'Forward (VRAM):     {(vram_forward - vram_0)/1e9:1.2f}GB, \n'
          f'Backward (VRAM):    {(th.cuda.max_memory_allocated(device) - vram_forward) / 1e9:1.2f}GB')

    return prediction, input.grad


if __name__ == "__main__":
    device = th.device(DEVICE)

    input = th.rand(size=SHAPE, device=device)
    label = th.rand_like(input)

    model = nn.Conv2d(
        in_channels=input.shape[1],
        out_channels=input.shape[1],
        kernel_size=(3, 3),
        padding=(1, 1),
        bias=True).to(device=device)

    criterion = nn.MSELoss()

    # For initializing device
    criterion(model(input), label).backward()

    print('New ReLU:')
    pr_new = speed_test(model, ReLU(), input, label, criterion, ITERATION)
    print('ReLU:')
    pr_relu = speed_test(model, nn.ReLU(), input, label, criterion, ITERATION)

    print('')
    print(f'Max abs difference prediction: {th.max((pr_new[0] - pr_relu[0]).abs()).detach().cpu():1.2e}')
    print(f'Max abs difference grad:       {th.max((pr_new[1] - pr_relu[1]).abs()).detach().cpu():1.2e}')

```

### Output:
New ReLU:
Forward + Backward: 0.50s, 
Forward (VRAM):     2.95GB, 
Backward (VRAM):    1.49GB

ReLU:
Forward + Backward: 0.50s, 
Forward (VRAM):     5.37GB, 
Backward (VRAM):    1.49GB

Max abs difference prediction: 0.00e+00
Max abs difference grad:       0.00e+00

**Comment by MengeTM** [comment]
LeakyReLU should work with something similar when using `grad *= LeakyReLU.negative_slope * (output < 0) + 1 * (output > 0)`

**Comment by albanD** [comment]
I agree with you :)
I feel like we should actually collapse the entries in derivatives.yaml from
```yaml
- name: relu(Tensor self) -> Tensor
  self: threshold_backward(grad, self, 0)

- name: relu_(Tensor(a!) self) -> Tensor(a!)
  self: threshold_backward(grad, result, 0)
```
to
```yaml
- name: relu(Tensor self) -> Tensor
  self: threshold_backward(grad, result, 0)
```

I am not sure though why it is not already done this way so we need to be careful in testing the impact.
Would you be able to send a PR making this change?

**Comment by MengeTM** [comment]
Using "result" instead of "self" results in a much larger VRAM need for both the proposed relu (my version above) and the native relu. How it is defined which tensors are saved for backward?

**Comment by MengeTM** [comment]
Saving "result" for leaky_relu works only for negative_slope >= 0.

**Comment by MengeTM** [comment]
Saving `result` for leaky_relu works only for `negative_slope >= 0`.

This results from input values < 0 becoming output values > 0 for a `negative_slope < 0`. Thus for `negative_slope=-1`, *result=5* may result from an *input=5* or an *input=-5*, which results in an ambiguous calculation of the gradients, so that `result` can not replace `self` for gradient calculations.

Possible Solutions:

### Do nothing
We may compute gradients for `leaky_relu()` with `self` as it was implemented before, which uses double the VRAM of other solutions

### Disallow negative_slope < 0
This is already done for `leaky_relu_()`, resulting in an Error for `negative_slope < 0`. However, this will potentially affect some users and change the behavior of `leaky_relu()`.

### Use self for gradients when negative_slope < 0
We may use `result` for `negative_slope >= 0`, resulting in a smaller VRAM need for most use-cases , and `self` for `negative_slope < 0`. However, while this is easy to be implemented with a `autograd.Function`, it seems to require a change of gen_autograd.py when implementing it for derivatives.yaml. This seems to be necessary, due to saved variables being determined at compile time.

``` Python
class LeakyReluFunction(Function):
    @staticmethod
    def forward(ctx, input: th.Tensor, negative_slope=1e-2) -> th.Tensor:
        relu = functional.leaky_relu(input, negative_slope)
        ctx.save_for_backward(relu if negative_slope >= 0 else input)
        ctx.negative_slope = negative_slope
        return relu

    @staticmethod
    def backward(ctx, grad: th.Tensor):
        relu, = ctx.saved_tensors
        negative_slope = ctx.negative_slope
        grad *= negative_slope * (relu <= 0) + (relu > 0)
        return grad, None

```

**Comment by MengeTM** [comment]
I opened a new issue #63644 for improving the needed VRAM of LeakyReLU and Elu, which need an improvement of the autograd generation.
